gamemodes/terrortown/gamemode/shared/sh_util.lua

```diff
@@ -107,27 +119,44 @@ end
 -- @realm shared
 function util.GetNextAlivePlayer(ply)
 	local alive = util.GetAlivePlayers()
-
 	if #alive < 1 then return end
 
-	local prev = nil
-	local choice = nil
-
 	if IsValid(ply) then
-		for _, p in pairs(alive) do
+		local prev = nil
+
+		for i = 1, #alive do
 			if prev == ply then
-				choice = p
+				return alive[i]
 			end
 
-			prev = p
+			prev = alive[i]
 		end
 	end
 
-	if not IsValid(choice) then
-		choice = alive[1]
-	end
+	return alive[1]
+end
+
+---
+-- Returns a random entry of the given @{table}
+-- @param table tbl the @{table} that contains the data
+-- @param nil|function filterFn the @{function} that has to return true on the given entry
+-- @return any the entry that returned true on the given @{function}
+-- @note The given @{table} has to be iterable.
+-- @note The returned entry will get removed from the given @{table}. If you wanna keep the original table untouched, create a copy for this function.
+-- @realm shared
+function table.GetRandomEntry(tbl, filterFn)
+	local entry = nil
+	local isInvalidFilterFn = not isfunction(filterFn)
+
+	repeat
+		if #tbl <= 0 then return end
+
+		local rnd = math.random(#tbl)
+		entry = tbl[rnd]
+		table.remove(tbl, rnd)
+	until isInvalidFilterFn or filterFn(entry)
```

Comment:
```suggestion
	local finalTbl = tbl
	
	if isfunction(filterFn) then
		local tmp = {}

		for i = 1, #tbl do
			if filterFn(tbl[i]) then
				tmp[#tmp + 1] = tbl[i]
			end
		end
		
		finalTbl = tmp
	end

	return #finalTbl > 0 and finalTbl[math.random(#finalTbl)] or nil
```

Would be the alternative.

Reply:
`return #finalTbl > 0 and finalTbl[math.random(#finalTbl)] or nil`

I'm still new to Lua, but how would this line handle the case where finalTbl has only one entry, which is zero or false (i.e. {0} or {false})? Wouldn't the condition return nil in this case (as opposed to 0 or false)?

I think something like this would be more readable in any case:
```
if #finalTbl > 0 then
    return finalTbl[math.random(#finalTbl)]
else
    return nil
end
```

Reply:
`finalTbl[math.random(#finalTbl)]` is a shortened code for `finalTbl[math.random(1, #finalTbl)]`. So if `#finalTbl` is `1`, then `math.random(1, 1)` is `1` (in 100%) too. So you are right so far.
Let's continue: if `finalTbl[1]` is `0` or `false`, then the function will return `0` or `false` (as intended). The function has no error-state or false-return in order to detect any issue. It's a simple getter method. So if you insert a list of numbers or booleans, it's intended that `0` or `false` will possibly return.

Just if the function returns `nil`, nothing was found.

---

Regarding your last code suggestion: Simple things should be kept simple. The `or nil` just is an addition to make it more readable. In your suggested code, you can cut the whole `else`-clause, without worrying about any changed mechanic. Running `return nil` or `return` at the end of a function isn't necessary.

so

```lua
if #finalTbl > 0 then
    return finalTbl[math.random(#finalTbl)]
end
```

could be an alternative. If writing lua-code on github, keep in mind to enable code highlighting (` ```lua [code] ``` `)

Reply:
@Alf21 I like the alternative more, but then I think we should make this a utility function for purely applying a filter on a table instead. So it can be reused in a more general use case and would separate the concerns better. (Your shortcut, on checking if a value is actually in a filter might also be realized, if you would want to keep this as a whole including the part of retrieving an item, by just checking the filter function result and early returning on true)